### ISSUE TITLE
StackBool is incorrect

### ISSUE THREAD

**Issue by durban** [scalaz/scalaz-zio#860]
`scalaz.zio.internal.StackBool` is incorrect (NB: a JVM long has 64 bits):

```scala
scala> import scalaz.zio.internal.StackBool
import scalaz.zio.internal.StackBool

scala> import java.util.Random
import java.util.Random

scala> val r = new Random(43L)
r: java.util.Random = java.util.Random@82fcb31

scala> val l = List.fill(65)(r.nextBoolean())
l: List[Boolean] = List(true, false, true, true, true, true, false, false, false, false, true, false, true, true, false, false, true, true, true, false, false, true, true, false, true, true, true, true, true, true, false, false, true, false, false, true, false, true, true, false, false, false, true, false, true, true, true, false, true, false, true, false, true, false, false, false, false, false, true, true, false, true, true, true, false)

scala> val s = StackBool()
s: scalaz.zio.internal.StackBool = StackBool()

scala> l.foreach(s.push)

scala> s.toList.reverse == l
res1: Boolean = false

```

**Comment by LGLO** [comment]
I think your observation is right and very important!

IMO code looks like Long has 32 bytes, not 8 bytes (64 bits) and what is more, I see that `getOrElse` doesn't work when stack has more that one Entry.

Would you like to provide a fix?

**Comment by vilu** [comment]
If not I would volunteer for this bug.

**Comment by LGLO** [comment]
@vilu 🙏 ❤️ 
I'm not that much fun of Scalacheck, but this looks like a valid place to introduce them.
Could you first remake tests to use random lists of Booleans?
I suggest to generate lists of size up to 64 - all tests should be still passing. When you change generators to bigger sizes, preferably at lest 129 - to employ 3 entries in Stack, tests should fail.

**Comment by NeQuissimus** [comment]
Or Scalaprops, I tend to prefer that...

**Comment by vilu** [comment]
I'll try to get some time to work on this this weekend.

**Comment by jdegoes** [comment]
@vilu Would love a fix for this!

**Comment by vilu** [comment]
Turned out I didn't have that much time this weekend. I have a working solution now but the code is a mess. I'll try to clean it up and create a wip PR for some feedback this weekend.

**Comment by vilu** [comment]
I've opened up a PR for some feedback: https://github.com/scalaz/scalaz-zio/pull/898

**Comment by vilu** [comment]
I went with Scalacheck because it was already on the classpath and I don't really have any preference.

**Comment by LGLO** [comment]
Closed by #898 
Thank you @vilu
